### PR TITLE
feat: dashboard statistics and get all boards with pagination

### LINKED ISSUE #123: [FEATURE]: Dashboard statistics and boards pagination
## Summary

- Method that count the users, teams and boards of the authenticated user
- Method that retrieve a list of boards depending of the user role  with pagination

## User story it relates to

#118 

### Motivation

We need this to show the user's boards in the dashboard and boards page

### PR DESCRIPTION
**Pull request by nunocaseiro**
Relates to #123

## Proposed Changes

  - Method in auth controller to retrieve some statistics about the authenticated user
  - Get all boards service method depending of the user's role and atual page in the frontend. 
     - If dashboard -> all boards of user of the last 3 months
     - if boards page and the user is super admin -> all boards 
     - if boards page and the user is not super admin -> all boards of user
 - Get all boards with pagination
 
This pull request is intended to close #123

### CODE REVIEW

File: backend/src/infrastructure/config/jwt.register.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { JwtModule } from '@nestjs/jwt';
+import {
+  JWT_ACCESS_TOKEN_SECRET,
+  JWT_ACCESS_TOKEN_EXPIRATION_TIME,
+} from '../../libs/constants/jwt';
+
+export const JwtRegister = JwtModule.registerAsync({
+  imports: [ConfigModule],
+  inject: [ConfigService],
+  useFactory: async (configService: ConfigService) => ({
```
Comment (rpvsilva) [inline backend/src/infrastructure/config/jwt.register.ts:11]:
```suggestion
  useFactory: (configService: ConfigService) => ({
```

File: backend/src/modules/auth/controller/auth.controller.ts [hunk 2]
```diff
@@ -84,4 +93,14 @@ export default class AuthController {
...
     return !!found;
   }
+
+  @UseGuards(JwtAuthenticationGuard)
+  @Get('/dashboardStatistics')
+  async getDashboardHeaderInfo(@Req() request: RequestWithUser) {
+    const { _id: userId } = request.user;
+    const usersCount = await this.getUserApp.countUsers();
+    const teamsCount = await this.getTeamsApp.countTeams(userId);
+    const boardsCount = await this.getBoardApp.countBoards(userId);
```
Comment (rpvsilva) [inline backend/src/modules/auth/controller/auth.controller.ts:103]:
```suggestion
    const [usersCount, teamsCount, boardsCount] = await Promise.all([
      this.getUserApp.countUsers(),
      this.getTeamsApp.countTeams(userId),
      this.getBoardApp.countBoards(userId),
    ]);
```

File: backend/src/modules/boards/controller/boards.controller.ts [hunk 3]
```diff
@@ -54,10 +56,25 @@ export default class BoardsController {
...
+
+  @UseGuards(JwtAuthenticationGuard)
+  @Get()
+  getAllBoards(
+    @Req() request: RequestWithUser,
+    @Query() { page, size }: PaginationParams,
+  ) {
+    const { _id: userId, isSAdmin } = request.user;
+    if (isSAdmin)
+      return this.getBoardApp.getBoards('allBoards', userId, page, size);
```
Comment (rpvsilva) [inline backend/src/modules/boards/controller/boards.controller.ts:76]:
```suggestion
    if (isSAdmin) {
      return this.getBoardApp.getBoards('allBoards', userId, page, size);
    }
```

File: backend/src/modules/boards/controller/boards.controller.ts [hunk 4]
```diff
@@ -54,10 +56,25 @@ export default class BoardsController {
...
-  @Get()
-  boards(@Req() request: RequestWithUser) {
+  @Get('/dashboard')
+  getDashboardBoards(
+    @Req() request: RequestWithUser,
+    @Query() { page, size }: PaginationParams,
+  ) {
     const { _id: userId } = request.user;
-    return this.getBoardApp.getAllBoards(userId);
+    return this.getBoardApp.getBoards('dashboard', userId, page, size);
```
Comment (rpvsilva) [inline backend/src/modules/boards/controller/boards.controller.ts:65]: You don't need to destructure the user object

File: backend/src/modules/boards/services/get.board.service.ts [hunk 5]
```diff
@@ -1,32 +1,163 @@
...
+      .find({ user: userId })
+      .select('board')
+      .distinct('board')
       .lean()
       .exec();
   }
 
+  async getAllBoardIdsAndTeamIdsOfUser(userId: string) {
+    const boardIds = await this.getAllBoardsIdsOfUser(userId);
+    const teamIds = await this.getTeamService.getTeamsOfUser(userId);
```
Comment (rpvsilva) [inline backend/src/modules/boards/services/get.board.service.ts:31]:
```suggestion
    const [boardIds, teamIds] = await Promise.all([
      this.getAllBoardsIdsOfUser(userId),
      this.getTeamService.getTeamsOfUser(userId)
    ])
```

File: backend/src/modules/boards/services/get.board.service.ts [hunk 6]
```diff
@@ -1,32 +1,163 @@
...
+    userId: string,
+  ) {
+    const { boardIds, teamIds } = await this.getAllBoardIdsAndTeamIdsOfUser(
+      userId,
+    );
+
+    if (option === 'dashboard') {
+      const last3Months = new Date(
+        new Date().setMonth(new Date().getMonth() - 3),
+      );
```
Comment (rpvsilva) [inline backend/src/modules/boards/services/get.board.service.ts:46]:
```suggestion
      const now = new Date();
      const last3Months = new Date(now.setMonth(now.getMonth() - 3));
```

File: backend/src/modules/boards/services/get.board.service.ts [hunk 7]
```diff
@@ -1,32 +1,163 @@
...
+    return {
+      $and: [
+        { isSubBoard: false },
+        { $or: [{ _id: { $in: boardIds } }, { team: { $ne: null } }] },
+      ],
+    };
+  }
+
+  async getBoards(
+    option: 'dashboard' | 'allBoards' | 'myBoards',
```
Comment (rpvsilva) [inline backend/src/modules/boards/services/get.board.service.ts:72]: I think this logic shouldn't be in server side, you just need to be concerned about the limit and offset sent by the client, the client is the one that needs to be concerned about what he wants to get

**Review by rpvsilva — CHANGES_REQUESTED** [review]

File: backend/src/modules/boards/services/get.board.service.ts [hunk 8]
```diff
@@ -1,32 +1,171 @@
...
+        select: 'user role -board',
+        populate: {
+          path: 'user',
+          select: 'firstName lastName',
+        },
+      })
+      .lean({ virtuals: true })
+      .exec();
+
+    return { boards, hasNextPage, page };
```
Comment (rpvsilva) [inline backend/src/modules/boards/services/get.board.service.ts:132]: here you are doing 2 requests each time to check if there is a next page, let's merge it and get a better solution for it

**Review by rpvsilva — APPROVED** [review]

File: backend/src/modules/boards/services/get.board.service.ts (hunk 8 above)
Comment (nunocaseiro) [reply backend/src/modules/boards/services/get.board.service.ts:132]: yes, I think we can do it with the aggregator method from mongo. We need to put more effort on it. Let's merge it then

**Review by rpvsilva — APPROVED** [review]

**Pull request merged**